cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go

```diff
@@ -1181,16 +1239,252 @@ func TestNodeGroupWithFailedMachine(t *testing.T) {
+		nodeGroupMinSizeAnnotationKey: "1",
+		nodeGroupMaxSizeAnnotationKey: "10",
+	}
+
+	type testCaseConfig struct {
+		nodeLabels         map[string]string
+		nodegroupLabels    map[string]string
+		includeNodes       bool
+		expectedErr        error
+		expectedCapacity   map[corev1.ResourceName]int64
+		expectedNodeLabels map[string]string
+	}
+
+	testCases := []struct {
+		name                 string
+		nodeGroupAnnotations map[string]string
+		config               testCaseConfig
+	}{
+		{
+			name: "When the NodeGroup cannot scale from zero",
+			config: testCaseConfig{
+				expectedErr: cloudprovider.ErrNotImplemented,
+			},
+		},
+		{
+			name: "When the NodeGroup can scale from zero",
+			nodeGroupAnnotations: map[string]string{
+				memoryKey:   "2048Mi",
+				cpuKey:      "2",
+				gpuTypeKey:  gpuapis.ResourceNvidiaGPU,
+				gpuCountKey: "1",
+			},
+			config: testCaseConfig{
+				expectedErr: nil,
+				expectedCapacity: map[corev1.ResourceName]int64{
+					corev1.ResourceCPU:        2,
+					corev1.ResourceMemory:     2048 * 1024 * 1024,
+					corev1.ResourcePods:       110,
+					gpuapis.ResourceNvidiaGPU: 1,
+				},
+				expectedNodeLabels: map[string]string{
+					"kubernetes.io/os":        "linux",
+					"beta.kubernetes.io/os":   "linux",
+					"kubernetes.io/arch":      "amd64",
+					"beta.kubernetes.io/arch": "amd64",
+				},
+			},
+		},
+		{
+			name: "When the NodeGroup can scale from zero and the nodegroup adds labels to the Node",
+			nodeGroupAnnotations: map[string]string{
+				memoryKey: "2048Mi",
+				cpuKey:    "2",
+			},
+			config: testCaseConfig{
+				expectedErr: nil,
+				nodegroupLabels: map[string]string{
+					"nodeGroupLabel": "value",
+					"anotherLabel":   "anotherValue",
+				},
+				expectedCapacity: map[corev1.ResourceName]int64{
+					corev1.ResourceCPU:    2,
+					corev1.ResourceMemory: 2048 * 1024 * 1024,
+					corev1.ResourcePods:   110,
+				},
+				expectedNodeLabels: map[string]string{
+					"kubernetes.io/os":        "linux",
+					"beta.kubernetes.io/os":   "linux",
+					"kubernetes.io/arch":      "amd64",
+					"beta.kubernetes.io/arch": "amd64",
+					"nodeGroupLabel":          "value",
+					"anotherLabel":            "anotherValue",
+				},
+			},
+		},
+		{
+			name: "When the NodeGroup can scale from zero and the Node still exists, it includes the known node labels",
+			nodeGroupAnnotations: map[string]string{
+				memoryKey: "2048Mi",
+				cpuKey:    "2",
+			},
+			config: testCaseConfig{
+				includeNodes: true,
+				expectedErr:  nil,
+				nodeLabels: map[string]string{
+					"kubernetes.io/os":                 "windows",
+					"kubernetes.io/arch":               "arm64",
+					"node.kubernetes.io/instance-type": "instance1",
+					"anotherLabel":                     "nodeValue", // This should not be copied as it is not a well known label
+				},
+				nodegroupLabels: map[string]string{
+					"nodeGroupLabel": "value",
+					"anotherLabel":   "anotherValue",
```

Comment:
Might be more obvious if this was as below, took me a moment to work out why it was another value and not the node value
```suggestion
					"anotherLabel":   "nodeGroupValue",
```